src/ripple/protocol/impl/STObject.cpp

```diff
@@ -603,11 +603,24 @@ void STObject::setFieldU32 (SField const& field, std::uint32_t v)
     setFieldUsingSetValue <STUInt32> (field, v);
 }
 
+void STObject::setFieldU32 (SField const& field, XRPAmount v)
+{
+    assert(v > beast::zero &&
+        v <= std::numeric_limits<std::uint32_t>::max());
+    setFieldUsingSetValue <STUInt32> (field, v.drops());
+}
+
 void STObject::setFieldU64 (SField const& field, std::uint64_t v)
 {
     setFieldUsingSetValue <STUInt64> (field, v);
 }
 
+void STObject::setFieldU64 (SField const& field, XRPAmount v)
+{
+    assert(v > beast::zero);
```

Comment:
I'm glad there's an assert here, but I'm thinking this might be a really bad error to let slip. Do we want to crash here instead?

Reply:
After some discussion, rather than range checking, we determined that it would likely be a lot safer to implement type safety. To that end, I templatized `XRPAmount`, leaving the "default" `XRPAmount` as `XRPAmountBase<std::int64_t>`, then rewrote the `setFieldU32` and `U64` to take `XRPAmountU32` and `U64`, respectively.

`XRPAmountBase` has an `explicit` ctor for other `XRPAmountBase`s with different template types, so they can be converted when necessary, but not accidentally.

I think this could work, yo.

Reply:
Update: No more converting ctors for these classes. Instead, there's an `as<Dest>()` function that does range checking if appropriate and returns the requested type.